crates/bevy_ecs/src/term_query/terms/component.rs

```diff
@@ -0,0 +1,549 @@
+    }
+
+    #[inline(always)]
+    unsafe fn from_fetch<'w>(_term: FetchedComponent<'w>) -> Self::Item<'w> {}
+}
+
+impl<T: Component> ComponentQueryTerm for Without<T> {
+    type Item<'w> = ();
+    type ReadOnly = Self;
+
+    fn init_term(world: &mut World) -> ComponentTerm {
+        let component = world.init_component::<T>();
+        ComponentTerm::without(component)
+    }
+
+    #[inline(always)]
+    unsafe fn from_fetch<'w>(_term: FetchedComponent<'w>) -> Self::Item<'w> {}
+}
+
+impl<T: Component> ComponentQueryTerm for Has<T> {
+    type Item<'w> = bool;
+    type ReadOnly = Self;
+
+    fn init_term(world: &mut World) -> ComponentTerm {
+        let component = world.init_component::<T>();
+        ComponentTerm::with(component).optional()
+    }
+
+    #[inline(always)]
+    unsafe fn from_fetch<'w>(term: FetchedComponent<'w>) -> Self::Item<'w> {
+        term.matched
+    }
+}
+
+impl<T: Component> ComponentQueryTerm for Added<T> {
+    type Item<'w> = ();
+    type ReadOnly = Self;
+
+    fn init_term(world: &mut World) -> ComponentTerm {
+        let component = world.init_component::<T>();
+        ComponentTerm::added(component)
+    }
+
+    #[inline(always)]
+    unsafe fn from_fetch<'w>(_term: FetchedComponent<'w>) -> Self::Item<'w> {}
+}
+
+impl<T: Component> ComponentQueryTerm for Changed<T> {
+    type Item<'w> = ();
+    type ReadOnly = Self;
+
+    fn init_term(world: &mut World) -> ComponentTerm {
+        let component = world.init_component::<T>();
+        ComponentTerm::changed(component)
+    }
+
+    #[inline(always)]
+    unsafe fn from_fetch<'w>(_term: FetchedComponent<'w>) -> Self::Item<'w> {}
+}
+
+impl<T: Component> ComponentQueryTerm for &T {
+    type Item<'w> = &'w T;
+    type ReadOnly = Self;
+
+    fn init_term(world: &mut World) -> ComponentTerm {
+        let component = world.init_component::<T>();
+        ComponentTerm::read_id(component)
+    }
+
+    #[inline(always)]
+    unsafe fn from_fetch<'w>(term: FetchedComponent<'w>) -> Self::Item<'w> {
+        term.pointer.debug_checked_unwrap().deref()
+    }
+}
+
+impl<T: Component> ComponentQueryTerm for Ref<'_, T> {
+    type Item<'w> = Ref<'w, T>;
+    type ReadOnly = Self;
+
+    fn init_term(world: &mut World) -> ComponentTerm {
+        let component = world.init_component::<T>();
+        ComponentTerm::read_id(component).with_change_detection()
+    }
+
+    #[inline(always)]
+    unsafe fn from_fetch<'w>(term: FetchedComponent<'w>) -> Self::Item<'w> {
+        let change_detection = term.change_ticks.as_ref().debug_checked_unwrap();
+        Ref {
+            value: term.pointer.debug_checked_unwrap().deref(),
+            ticks: Ticks {
+                added: change_detection.added.deref(),
+                changed: change_detection.changed.deref(),
+
+                last_run: change_detection.last_run,
+                this_run: change_detection.this_run,
+            },
+        }
+    }
+}
+
+impl ComponentQueryTerm for Ptr<'_> {
```

Comment:
Can you explain this one a bit more? We don't have this for static queries: why do we need it for dynamic ones? How does it work?

Reply:
Yes, this is modelled after the existing work in #6390. It's intended to be used when you want access to a component at runtime but you don't know it's type. You can set this as one of the terms in your iterator and then access the data directly or cast it there.

Here's a test that does this:
```rust
    #[test]
    fn builder_ptr_dynamic() {
        let mut world = World::new();
        let entity = world.spawn((A(0), B(1))).id();
        let component_id_a = world.init_component::<A>();
        let component_id_b = world.init_component::<B>();

        // Using term_at is currently unsafe as it allows you to edit the targets of arbitrary terms
        // possibly putting the terms in the iterator out of sync with the internal state
        let mut query = unsafe {
            QueryBuilder::<(Entity, Ptr, Ptr)>::new(&mut world)
                .term_at(1)
                .set_id(component_id_a)
                .term_at(2)
                .set_id(component_id_b)
                .build()
        };

        let (e, a, b) = query.single(&world);

        assert_eq!(e, entity);

        let a = unsafe { a.deref::<A>() };
        let b = unsafe { b.deref::<B>() };

        assert_eq!(0, a.0);
        assert_eq!(1, b.0);
    }
```

Reply:
Cool! Okay, we should *definitely* try to document that somewhere as this PR is polished.